src/PagesDisplaySection.php

```diff
@@ -1,61 +1,73 @@
 <?php
 
 use Kirby\Exception\InvalidArgumentException;
-use Kirby\Toolkit\Str;
-use Kirby\Toolkit\Query;
 use Kirby\Cms\Section;
+use Kirby\Toolkit\Query;
 
 $base = Section::$types['pages'];
-
-return array_replace_recursive($base, [
+$extension = [
     'props' => [
         'sortable' => function (bool $sortable = true) {
             return false;
         },
-        'query' => function(string $query = 'page.children') {
+        'query' => function (string $query = 'page.children') {
             return $query;
         }
     ],
     'computed' => [
-       'pages' => function () {
+        'pages' => function () {
+            $kirby = kirby();
             $q = new Query($this->query, [
-                'site' => site(),
-                'page' => $this->model(),
-                'pages' => site()->pages(),
-                'kirby' => kirby()
+                'kirby' => $kirby,
+                'site' => $kirby->site(),
+                'pages' => $kirby->site()->pages(),
+                'page' => $this->model()
             ]);
 
             $pages = $q->result();
 
-            if(is_a($pages, 'Kirby\\Cms\\Pages')) {
-                // show only readable
-                $pages = $pages->filterBy("isReadable", true);
-
-                // sort
-                if ($this->sortBy) {
-                    $pages = $pages->sortBy(...Str::split($this->sortBy, ' '));
-                }
-                
-                // pagination
-                $pages = $pages->paginate([
-                    'page' => $this->page,
-                    'limit' => $this->limit
-                ]);
-
-                return $pages;
-            } else {
+            if (!is_a($pages, \Kirby\Cms\Pages::class)) {
+                $result = $pages === null ? 'null' : get_class($pages);
                 throw new InvalidArgumentException(
-                    'Invalid query result - Result must be of type Kirby\\Cms\\Pages, ' 
-                    . ($pages === NULL ? 'NULL' : get_class($pages))
-                    . ' given.'
+                    "Query result must be of type \"Kirby\\Cms\\Pages\", \"{$result}\" given"
                 );
             }
+
+            // Loop for the best performance
+            foreach ($pages->data as $id => $page) {
+                // Remove all protected pages
+                if (!$page->isReadable()) {
+                    unset($pages->data[$id]);
+                    continue;
+                }
+
+                // Filter by all set templates
+                if ($this->templates && !in_array($page->intendedTemplate()->name(), $this->templates)) {
+                    unset($pages->data[$id]);
+                    continue;
+                }
+            }
+
+            // Sort pages
+            if ($this->sortBy) {
+                $pages = $pages->sort(...$pages::sortArgs($this->sortBy));
+            }
+
```

Comment:
Could you add the "flip" option, since we're already here? 🙏

Reply:
Good point. Sure! Done. (https://github.com/rasteiner/k3-pagesdisplay-section/pull/7)